Go stateless

davidaurelio opened the pull request:
- base dispatcher implementation
- stateful dispatcher wrapper for browsers
- static dispatcher wrapper for servers

davidaurelio committed:
Setup

davidaurelio committed:
Start dispatcher

davidaurelio committed:
Remove specs and code

davidaurelio committed:
Add iteration code

davidaurelio committed:
Break out of promise stacks to prevent errors from being silenced

davidaurelio committed:
Switch to pinkie-promise – es6-promise pollutes the global object

davidaurelio committed:
Fix iteration

davidaurelio committed:
Work on dispatcher

davidaurelio committed:
Start to change contract + implementation

davidaurelio committed:
Remove iter module

davidaurelio committed:
Add iterator creation function

davidaurelio committed:
Push changes…

Commit:
Fix function name

Commit:
Handle additional actions returned by stores

Commit:
Make dispatcher handle promises as expected

Commit:
Use `equals` to optimize callback invocation

Commit:
add object access functions for immutable.js

Commit:
Add `DataPointer` to be able to be able to share data across dispatches

Commit:
Add a stateful dispatcher that shares data across dispatches

Commit:
Add more tests to ensure interaction of the dispatcher with the data pointer

Commit:
Adapt stateful dispatcher

Commit:
Add static dispatcher

klipstein commented on dispatcher/index-test.js:
```diff
@@ -0,0 +1,568 @@
+      data = {a: {}, b: [null, null, {c: {}}]};
+    });
+
+    it('passes subtrees of the data to handlers, depending on registration', function() {
```
Interesting feature. Wondering what benefits you get from this in comparison to forwarding the full data tree?

davidaurelio replied:
- potentially more reusable stores
- potentially pluggable sub-components
- shallower trees to handle in stores
- if you subscribe stores on subtrees, and want to change the shape of the data root tree, you don’t need to refactor the stores.
- since you can subscribe to multiple sub-trees, you can reach into the specific data you depend on beforehand, that means, the store does not need to do the lookup itself, but will get the data it depends on passed in directly.

klipstein replied:
Understood it. So the register part is tightly coupled to the contract of the store then. This should be documented properly then.

davidaurelio committed:
Get rid of `DataPointer` in favor of an optional getter function.

- Makes the signatures of the base dispatcher and of wrappers more compatible
- Feels more natural than wrapping data

Commit:
Expose a ready-made dispatcher for Immutable.js

Commit:
Add change log

Commit:
Add helper to bind action creators to stateful dispatcher

Commit:
0.2.0-alpha.2

Commit:
Fix setter for immutable.js

Commit:
0.2.0-alpha.3

klipstein commented on dispatcher/immutable.js:
```diff
@@ -0,0 +1,23 @@
+var Dispatcher = require('.');
+
+function get(object, keypath) {
+  return object.getIn(keypath);
```
Is that protocol supported in every immutable implementation or is that specific to immutable.js?

davidaurelio replied:
it is specific to immutable.js

klipstein replied:
Then document it for DispatcherForImmutable or call it DispatcherForImmutableJs

klipstein commented on dispatcher/immutable.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+function set(object, keypath, value) {
+  return object.setIn(keypath, value);
```
Is that protocol supported in every immutable implementation or is that specific to immutable.js?

davidaurelio replied:
same here

klipstein commented on dispatcher/static.js:
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+module.exports = function StaticDispatcher(dispatcher) {
```
It would be nice to have a one-liner that describes for what a static dispatcher could be used.

klipstein commented on dispatcher/stateful.js:
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+module.exports = function StatefulDispatcher(dispatcher, data, onData) {
```
It would be nice to have a one-liner that describes for what a static dispatcher could be used.

davidaurelio replied:
Don’t copy and paste ;-) This is the _stateful_ dispatcher.

Would you add a comment in the source or just document it?

basecode replied:
+1 I also wonder when I should use a `StaticDispatcher` and when a `StatefulDispatcher`.

klipstein replied:
Oh. Copy+Paste mistake. Just wanted to highlight that a short description would help.

davidaurelio replied:
yes, I figured that.

My impression was, that `BrowserDispatcher` and `ServerDispatcher` are awful names.

So, the `StatefulDispatcher` wraps state and retains the root data tree over time, notifying its client about each data change via the callback.

The `StaticDispatcher` on the other hand does not wrap or retain data across dispatches, and will call the callback only once, as soon as no further actions are pending. I am not so convinced about that name, though. Better suggestions?

klipstein replied:
What about `StatelessDispatcher` and `StatefulDispatcher` and an additional doc string that explains that a `StatelessDispachter` can be used for stateless server rendering and the `StatefulDispatcher` is useful for a stateful client?

davidaurelio replied:
That’s the obvious choice, of course, but the base dispatcher is stateless, too. What makes the static dispatcher special is that it will only call the callback passed to `dispatch` once.

klipstein replied:
`OnceDispatcher` :)

klipstein commented on dispatcher/index.js:
```diff
@@ -0,0 +1,155 @@
+    },
+
+    dispatch: function(action, initialData, callback, updatedData) {
+      dispatch([action], [], initialData, updatedData, callback);
```
Please also change order here to: `... callback, updatedData`.

klipstein replied:
The name `updatedData` is kind of confusing. `getUpdatedData` would be a more obvious name for it.

davidaurelio replied:
Yes, I was just influenced by the “getters imply side effects” discussion we’ve had with Wolfram.

klipstein replied:
Then you should be consistent with that: https://github.com/uxebu/fluxebu/pull/2/files#diff-371cfd10f3cff30567b9da389111fc14R9

klipstein commented on dispatcher/index.js:
```diff
@@ -0,0 +1,155 @@
+    );
+  }
+
+  function dispatch(actionQueue, pendingPromises, data, updatedData, callback) {
```
Also here: `getUpdatedData` instead of `updatedData`.

klipstein commented on package.json:
```diff
@@ -0,0 +1,36 @@
+    "flux",
+    "server",
+    "browser",
+    "isomorphic",
```
universal :)

davidaurelio replied:
right
